gcovr/writer/cobertura.py

```diff
@@ -74,14 +74,52 @@ def print_cobertura_report(covdata: CovData, output_file, options):
             ),
         )
         c = etree.Element("class")
-        # The Cobertura DTD requires a methods section, which isn't
-        # trivial to get from gcov (so we will leave it blank)
-        etree.SubElement(c, "methods")
-        lines = etree.SubElement(c, "lines")
 
+        sorted_lines = sorted(data.lines)
+        iter_reversed_sorted_lines = iter(reversed(sorted_lines))
+        methods = []
+        for function in sorted(data.functions, key=lambda f: data.functions[f].lineno, reverse=True):
+            function_cov = data.functions[function]
+            elem = _method_element(function_cov)
+            methods.append(elem)
+
+            lines = etree.Element("lines")
+            elem.append(lines)
+            function_branch = CoverageStat(0, 0)
+            function_linenos = []
+            for lineno in iter_reversed_sorted_lines:
+                if lineno >= function_cov.lineno:
+                    function_linenos.insert(0, lineno)
+                else:
+                    break
+            # Dummy file to get line coverage for function
+            dummy_file_coverage = FileCoverage("dummy")
+            for lineno in function_linenos:
+                line_cov = data.lines[lineno]
+                dummy_file_coverage.lines[lineno] = line_cov
+                if not (line_cov.is_covered or line_cov.is_uncovered):
+                    continue
+
+                b = line_cov.branch_coverage()
+                if b.total:
+                    function_branch += b
+                lines.append(_line_element(line_cov))
```

Comment:
This looks like duplicate code to the line element generation below – could probably extracted into a function.